File: app/src/main/java/com/saienko/androidthings/barman/ui/dialog/CustomDialog.java
package com.saienko.androidthings.barman.ui.dialog;

import android.app.Activity;
import android.content.Context;
import android.content.DialogInterface;
import android.support.v7.app.AlertDialog;
import android.view.ContextThemeWrapper;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Spinner;
import com.saienko.androidthings.barman.R;
import com.saienko.androidthings.barman.db.cocktail.CocktailElement;
import com.saienko.androidthings.barman.db.cocktail.Component;
import com.saienko.androidthings.barman.db.drinkGroup.CocktailGroup;

import java.util.List;

/**
 * Created
 * User: Vasiliy Saienko
 * Date: 9/9/17
 * Time: 16:39
 */

public class CustomDialog {

    public static AlertDialog showDialog(Context context, String title, String message,
                                         final DialogInterface.OnClickListener positiveListener,
                                         final DialogInterface.OnClickListener negativeListener) {

        final AlertDialog ad = new AlertDialog.Builder(context)
                .setTitle(title)
                .setMessage(message)
                .setPositiveButton(android.R.string.ok, positiveListener)
                .setNegativeButton(android.R.string.cancel, negativeListener)
                .create();
        ad.show();
        ad.setCanceledOnTouchOutside(false);
        ad.setCancelable(false);
        return ad;
    }

    public static AlertDialog showDialog(Context context, String title, String message,
                                         final DialogInterface.OnClickListener positiveListener) {

        final AlertDialog ad = new AlertDialog.Builder(context)
                .setTitle(title)
                .setMessage(message)
                .setPositiveButton(android.R.string.ok, positiveListener)
                .setNegativeButton(android.R.string.cancel, (dialogInterface, i) -> {})
                .create();
        ad.show();
        ad.setCanceledOnTouchOutside(false);
        ad.setCancelable(false);
        return ad;
    }

    public static AlertDialog addNewItem(Context context, String title, String message,
                                         String text,
                                         final OnSaveTextListener callback) {
        final EditText edittext = new EditText(new ContextThemeWrapper(context, R.style.EditText_Dialog), null, 0);
        edittext.setText(text);
        final AlertDialog ad = new AlertDialog.Builder(context)
                .setTitle(title)
                .setMessage(message)
                .setPositiveButton("Save", (dialog, which) -> callback.onTextSaved(edittext.getText().toString()))
                .setNegativeButton(android.R.string.cancel, (dialogInterface, i) -> {})
                .setView(edittext)
                .create();
        ad.show();
        return ad;
    }

    public static AlertDialog addNewDrink(Context context, String title, String message,
                                          List<CocktailGroup> list,
                                          String text,
                                          String positiveText,
                                          final OnSaveTextListener callback) {
        final EditText edittext =
                new EditText(new ContextThemeWrapper(context, R.style.EditText_Dialog), null, 0);
        DrinkGroupDialogAdapter adapter = new DrinkGroupDialogAdapter(context, list);
        edittext.setText(text);
        final AlertDialog ad = new AlertDialog.Builder(context)
                .setTitle(title)
                .setMessage(message)
                .setAdapter(adapter, null)
                .setPositiveButton(positiveText, (dialog, which) -> callback.onTextSaved(edittext.getText().toString()))
                .setNegativeButton(android.R.string.cancel, (dialogInterface, i) -> {})
                .setView(edittext)
                .create();
        ad.show();
        return ad;
    }

    public static void editCocktailElement(Activity activity, CocktailElement cocktailElement,
                                           final OnAddComponentListener callback) {
        int            volume   = cocktailElement.getVolume();
        LayoutInflater inflater = activity.getLayoutInflater();

        AlertDialog.Builder builder    = new AlertDialog.Builder(activity);
        View                view       = inflater.inflate(R.layout.dialog_edit_component, null);
        ImageButton         ibMinus    = view.findViewById(R.id.ibMinus);
        ImageButton         ibPlus     = view.findViewById(R.id.ibPlus);
        ImageButton         ibMinusTen = view.findViewById(R.id.ibMinusTen);
        ImageButton         ibPlusTen  = view.findViewById(R.id.ibPlusTen);
        EditText            etVolume   = view.findViewById(R.id.etVolume);
        etVolume.setText(String.valueOf(volume));
        ibMinus.setOnClickListener(view12 -> {
            int value = Integer.valueOf(etVolume.getText().toString());
            if (value > 2) {
                value--;
                etVolume.setText(String.valueOf(value));
            }
        });
        ibMinusTen.setOnClickListener(view13 -> {
            int value = Integer.valueOf(etVolume.getText().toString());
            if (value > 11) {
                value = volume - 10;
                etVolume.setText(String.valueOf(value));
            }
        });


        ibPlus.setOnClickListener(view1 -> {
            int value = Integer.valueOf(etVolume.getText().toString());
            value++;
            etVolume.setText(String.valueOf(value));
        });
        ibPlusTen.setOnClickListener(view1 -> {
            int value = Integer.valueOf(etVolume.getText().toString());
            value = value + 10;
            etVolume.setText(String.valueOf(value));
        });

        builder.setView(view)
               .setTitle("Edit cocktail Component")
               .setPositiveButton("Save",
                                  (dialog, which) -> callback
                                          .onComponentEdit(cocktailElement,
                                                           Integer.valueOf(etVolume.getText().toString())))
               .setNegativeButton(android.R.string.cancel, (dialogInterface, i) -> {})
               .create();
        builder.show();
    }

    public static void addNewCocktailElement(Activity activity,
                                             List<Component> list,
                                             final OnAddComponentListener callback) {

        int                    volume   = 50;
        ComponentDialogAdapter adapter  = new ComponentDialogAdapter(activity, list);
        LayoutInflater         inflater = activity.getLayoutInflater();

        AlertDialog.Builder builder          = new AlertDialog.Builder(activity);
        View                view             = inflater.inflate(R.layout.dialog_add_component, null);
        Spinner             componentSpinner = view.findViewById(R.id.componentSpinner);
        ImageButton         ibMinus          = view.findViewById(R.id.ibMinus);
        ImageButton         ibPlus           = view.findViewById(R.id.ibPlus);
        ImageButton         ibMinusTen       = view.findViewById(R.id.ibMinusTen);
        ImageButton         ibPlusTen        = view.findViewById(R.id.ibPlusTen);

        EditText          etVolume          = view.findViewById(R.id.etVolume);
        final Component[] selectedComponent = new Component[1];
        componentSpinner.setAdapter(adapter);
        componentSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                selectedComponent[0] = list.get(i);
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {
                // IGNORE
            }
        });
        ibMinus.setOnClickListener(view12 -> {
            int value = Integer.valueOf(etVolume.getText().toString());
            if (value > 2) {
                value--;
                etVolume.setText(String.valueOf(value));
            }
        });
        ibMinusTen.setOnClickListener(view13 -> {
            int value = Integer.valueOf(etVolume.getText().toString());
            if (value > 11) {
                value = volume - 10;
                etVolume.setText(String.valueOf(value));
            }
        });
        ibPlus.setOnClickListener(view1 -> {
            int value = Integer.valueOf(etVolume.getText().toString());
            value++;
            etVolume.setText(String.valueOf(value));
        });
        ibPlusTen.setOnClickListener(view1 -> {
            int value = Integer.valueOf(etVolume.getText().toString());
            value = value + 10;
            etVolume.setText(String.valueOf(value));
        });

        builder.setView(view)
               .setTitle("Add new cocktail Component")
               .setPositiveButton("Save",
                                  (dialog, which) -> callback
                                          .onComponentAdded(selectedComponent[0],
                                                            Integer.valueOf(etVolume.getText().toString())))
               .setNegativeButton(android.R.string.cancel, (dialogInterface, i) -> {})
               .create();
        builder.show();
    }
}

File: app/src/main/java/com/saienko/androidthings/barman/ui/dialog/CustomDialog.kt
package com.saienko.androidthings.barman.ui.dialog

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.view.ContextThemeWrapper
import android.widget.EditText
import com.saienko.androidthings.barman.R


/**
 * Created
 * User: Vasiliy Saienko
 * Date: 9/9/17
 * Time: 16:39
 */

object CustomDialog {
//
//    fun showDialog(context: Context, title: String, message: String,
//                   positiveListener: DialogInterface.OnClickListener,
//                   negativeListener: DialogInterface.OnClickListener): AlertDialog {
//
//        val ad = AlertDialog.Builder(context)
//                .setTitle(title)
//                .setMessage(message)
//                .setPositiveButton(android.R.string.ok, positiveListener)
//                .setNegativeButton(android.R.string.cancel, negativeListener)
//                .create()
//        ad.show()
//        ad.setCanceledOnTouchOutside(false)
//        ad.setCancelable(false)
//        return ad
//    }

    fun showDialog(context: Context, title: String, message: String,
                   positiveListener: DialogInterface.OnClickListener): AlertDialog {

        return AlertDialog.Builder(context)
                .setTitle(title)
                .setMessage(message)
                .setPositiveButton(android.R.string.ok, positiveListener)
                .setNegativeButton(android.R.string.cancel) { _, _ -> }
                .create()
                .apply {
                    show()
                    setCanceledOnTouchOutside(false)
                    setCancelable(false)
                }
    }

    fun selectOneOfTwo(context: Context, title: String,
                       message: String,
                       firstButtonText: String,
                       secondButtonText: String,
                       firstAction: ActionListener,
                       secondAction: ActionListener) {
        val builder = AlertDialog.Builder(context)
        builder.setTitle(title)
        builder.setMessage(message)

        builder.setPositiveButton(firstButtonText) { dialog, _ ->
            firstAction.onAction()
            dialog.dismiss()
        }

        builder.setNegativeButton(secondButtonText) { dialog, _ ->
            secondAction.onAction()
            dialog.dismiss()
        }

        builder.setNeutralButton(context.resources.getString(android.R.string.cancel)) { dialog, _ ->
            dialog.dismiss()
        }
        val alertDialog = builder.create()
        alertDialog.show()
    }

    fun addNewItem(context: Context, title: String, message: String,
                   text: String,
                   callback: OnSaveTextListener): AlertDialog {
        val edittext = EditText(ContextThemeWrapper(context, R.style.EditText_Dialog), null, 0)
        edittext.setText(text)
        val ad = AlertDialog.Builder(context)
                .setTitle(title)
                .setMessage(message)
                .setPositiveButton("Save") { _, _ -> callback.onTextSaved(edittext.text.toString(), 1.0) }
                .setNegativeButton(android.R.string.cancel) { _, _ -> }
                .setView(edittext)
                .create()
        ad.show()
        return ad
    }

//    fun editCocktailElement(activity: Activity, cocktailElement: CocktailElement,
//                            callback: OnAddComponentListener) {
//        val volume = cocktailElement.volume
//        val inflater = activity.layoutInflater
//
//        val builder = AlertDialog.Builder(activity)
//        val view = inflater.inflate(R.layout.dialog_edit_component, null)
//        val ibMinus = view.findViewById<ImageButton>(R.id.ibMinus)
//        val ibPlus = view.findViewById<ImageButton>(R.id.ibPlus)
//        val ibMinusTen = view.findViewById<ImageButton>(R.id.ibMinusTen)
//        val ibPlusTen = view.findViewById<ImageButton>(R.id.ibPlusTen)
//        val etVolume = view.findViewById<EditText>(R.id.etVolume)
//
//        etVolume.setText(volume.toString())
//        ibMinus.setOnClickListener {
//            var value = Integer.valueOf(etVolume.text.toString())
//            if (value > 2) {
//                value--
//                etVolume.setText(value.toString())
//            }
//        }
//        ibMinusTen.setOnClickListener {
//            var value = Integer.valueOf(etVolume.text.toString())
//            if (value > 11) {
//                value -= 10
//                etVolume.setText(value.toString())
//            }
//        }
//
//
//        ibPlus.setOnClickListener {
//            var value = Integer.valueOf(etVolume.text.toString())
//            value++
//            etVolume.setText(value.toString())
//        }
//        ibPlusTen.setOnClickListener {
//            var value = Integer.valueOf(etVolume.text.toString())
//            value += 10
//            etVolume.setText(value.toString())
//        }
//
//        builder.setView(view)
//                .setTitle("Edit cocktail Component")
//                .setPositiveButton("Save"
//                                  ) { _, _ ->
//                    callback
//                            .onComponentEdit(cocktailElement,
//                                             Integer.valueOf(etVolume.text.toString()))
//                }
//                .setNegativeButton(android.R.string.cancel) { _, _ -> }
//                .create()
//        builder.show()
//    }

//    fun addNewCocktailElement(activity: Activity,
//                              list: List<Component>,
//                              callback: OnAddComponentListener) {
//
//        val adapter = ComponentDialogAdapter(activity, list)
//        val inflater = activity.layoutInflater
//
//        val builder = AlertDialog.Builder(activity)
//        val view = inflater.inflate(R.layout.dialog_add_component, null)
//        val componentSpinner = view.findViewById<Spinner>(R.id.componentSpinner)
//        val ibMinus = view.findViewById<ImageButton>(R.id.ibMinus)
//        val ibPlus = view.findViewById<ImageButton>(R.id.ibPlus)
//        val ibMinusTen = view.findViewById<ImageButton>(R.id.ibMinusTen)
//        val ibPlusTen = view.findViewById<ImageButton>(R.id.ibPlusTen)
//
//        val etVolume = view.findViewById<EditText>(R.id.etVolume)
//        var selectedComponent: Component? = null
//        componentSpinner.adapter = adapter
//        componentSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
//
//            override fun onItemSelected(adapterView: AdapterView<*>, view: View, i: Int, l: Long) {
//                selectedComponent = list[i]
//            }
//
//            override fun onNothingSelected(adapterView: AdapterView<*>) {
//                // IGNORE
//            }
//        }
//        ibMinus.setOnClickListener {
//            var value = Integer.valueOf(etVolume.text.toString())
//            if (value > 2) {
//                value--
//                etVolume.setText(value.toString())
//            }
//        }
//        ibMinusTen.setOnClickListener {
//            var value = Integer.valueOf(etVolume.text.toString())
//            if (value > 11) {
//                value -= 10
//                etVolume.setText(value.toString())
//            }
//        }
//        ibPlus.setOnClickListener {
//            var value = Integer.valueOf(etVolume.text.toString())
//            value++
//            etVolume.setText(value.toString())
//        }
//        ibPlusTen.setOnClickListener {
//            var value = Integer.valueOf(etVolume.text.toString())
//            value += 10
//            etVolume.setText(value.toString())
//        }
//
//        builder.setView(view)
//                .setTitle("Add new cocktail Component")
//                .setPositiveButton("Save"
//                                  ) { _, _ ->
//                    callback
//                            .onComponentAdded(selectedComponent,
//                                              Integer.valueOf(etVolume.text.toString()))
//                }
//                .setNegativeButton(android.R.string.cancel) { _, _ -> }
//                .create()
//        builder.show()
//    }
}

interface ActionListener {
    fun onAction()
}
